### PR TITLE
WIP: Merge Release-2.0 changes into master

### PR DESCRIPTION
**Pull request by jrising**

### CODE REVIEW

File: openest/generate/functions.py [hunk 1]
```diff
@@ -927,16 +934,20 @@ class AuxiliaryResultApplication(calculation.Application):
...
+    def __init__(self, region, subapp_main, subapp_aux, keeplastonly):
         super(AuxiliaryResultApplication, self).__init__(region)
         self.subapp_main = subapp_main
         self.subapp_aux = subapp_aux
+        self.keeplastonly = keeplastonly
 
     def push(self, ds):
         for yearresult in self.subapp_main.push(ds):
             for yearresult_aux in self.subapp_aux.push(ds):
                 next # Just take the last one
```
Comment (brews) [inline openest/generate/functions.py:946]: Not strictly in your PR but this `next` statement wont do anything. Needs to be `next()`?

File: openest/generate/smart_curve.py [hunk 2]
```diff
@@ -10,6 +10,7 @@
 """
 
 import numpy as np
+import xarray as xr
```
Comment (brews) [inline openest/generate/smart_curve.py:13]: I don't think this is used anywhere. Could just remove?

File: openest/generate/smart_curve.py [hunk 3]
```diff
@@ -351,22 +356,30 @@ def __init__(self, curve, offset):
...
-    def get_univariate(self):
-        return curve.ClippedCurve(self.curve.get_univariate(), self.cliplow)
+    @property
+    def univariate(self):
+        return curve.ClippedCurve(self.curve.univariate, self.cliplow)
+
+class OtherClippedCurve(curve.OtherClippedCurve, SmartCurve):
+    @property
+    def univariate(self):
+        return curve.ClippedCurve(self.clipping_curve.univariate, self.curve.univariate, self.clipy)
```
Comment (brews) [inline openest/generate/smart_curve.py:378]: There is an extra arg to `curve.ClippedCurve`?

File: test/curves/test_ushape.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+# Read polymins
+polymins = pd.read_csv("polymins.csv")
+allcalcs = pd.read_csv("allcalcs.csv")
+
+tas = np.arange(-40, 50, .01)
+fillins = np.arange(-40, 50, .1)
+    
+for ii in range(polymins.shape[0]):
+    print ii
```
Comment (brews) [inline test/curves/test_ushape.py:14]: I imagine this was aiming for `print(ii)`.

File: test/curves/test_ushape_marginal.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import pandas as pd
+from openest.curves.basic import ZeroInterceptPolynomialCurve, ClippedCurve, ShiftedCurve, MinimumCurve
+from openest.curves import ushape_numeric, ushape_analytic
+
+# Read polymins
+polymins = pd.read_csv("polymins.csv")
+allcalcs = pd.read_csv("allcalcs.csv")
+
+for ii in range(polymins.shape[0]):
+    print ii
```
Comment (brews) [inline test/curves/test_ushape_marginal.py:11]: This won't work.

File: openest/generate/functions.py (hunk 1 above)
Comment (jrising) [reply openest/generate/functions.py:946]: I believe it is written correctly. The goal here is to do nothing within this inner for loop, just allowing it to run.

File: openest/generate/fast_dataset.py [hunk 6]
```diff
@@ -461,6 +464,40 @@ def assert_index_equal(one, two):
...
+                    newvars[var] = (dims, ds[var].values[:, indices])
+                else:
+                    newvars[var] = (dims, ds[var].values[indices, :])
+            else:
+                tindex = list(dims).index(dim)
+                allindices = [slice(None)] * len(dims)
+                allindices[tindex] = indices
+                newvars[var] = (dims, ds[var].values[tuple(allindices)])
+        except Exception as ex:
+            print(("Failed to reorder %s for %s" % (var, self.reader)))
```
[the remark this answers is not in the document]
Comment (jrising) [reply openest/generate/fast_dataset.py]: Fixed.

File: openest/generate/fast_dataset.py [hunk 7]
```diff
@@ -461,6 +464,40 @@ def assert_index_equal(one, two):
...
+                if dims.index(dim) == 0:
+                    newvars[var] = (dims, ds[var].values[:, indices])
+                else:
+                    newvars[var] = (dims, ds[var].values[indices, :])
+            else:
+                tindex = list(dims).index(dim)
+                allindices = [slice(None)] * len(dims)
+                allindices[tindex] = indices
+                newvars[var] = (dims, ds[var].values[tuple(allindices)])
+        except Exception as ex:
```
[the remark this answers is not in the document]
Comment (jrising) [reply openest/generate/fast_dataset.py]: Dropped `as ex`.

File: openest/generate/smart_curve.py (hunk 2 above)
Comment (jrising) [reply openest/generate/smart_curve.py]: Dropped.

File: openest/generate/smart_curve.py (hunk 3 above)
Comment (jrising) [reply openest/generate/smart_curve.py]: Should have been `OtherClippedCurve`.

File: test/curves/test_ushape.py (hunk 4 above)
Comment (jrising) [reply test/curves/test_ushape.py]: Fixed.

File: test/curves/test_ushape_marginal.py (hunk 5 above)
Comment (jrising) [reply test/curves/test_ushape_marginal.py]: Fixed.

File: openest/generate/functions.py (hunk 1 above)
Comment (brews) [reply openest/generate/functions.py:946]: I guess I've never seen `next` as a statement and I can't find it documented in core python docs but it seems to work so 👍 . Nevermind!

Marking as resolved.

Comment (jrising) [reply openest/generate/functions.py]: Actually, the python statement I wanted is called `pass`. `next` was working for the wrong reasons. I've swapped it over.

**Comment by jrising** [comment]
@brews Github says there are still changes requested, but I think I've addressed your items. In any case, is there anything else you see that I should do before merging?

Thanks for going through all of it!

**Review by brews — APPROVED** [review]
Hey @jrising, that's great thank you.

I checked the change against downstream system tests and everything passes. So, 👍 

At this point my only comment is that many of the open-estimate unit tests do not run, or fail. CI is not running some tests. Some tests are failing to even collect because of missing resources, etc. Because this PR has big changes, I suggest we merge it and then do smaller PRs/issues to clean things up on a case-by-case basis.

Thank you!

**Comment by jrising** [comment]
Okay, great. Thanks for checking it!

**Pull request merged**